File: test/unit/Geometry/GMUGeoJSONParserTest.m
/* Copyright (c) 2016 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#import <XCTest/XCTest.h>

#import "GMUFeature.h"
#import "GMUGeometryCollection.h"
#import "GMULineString.h"
#import "GMUPoint.h"
#import "GMUPolygon.h"

#import "GMUGeoJSONParser.h"

@interface GMUGeoJSONParserTest : XCTestCase
@end

@implementation GMUGeoJSONParserTest

- (NSArray<GMUFeature *> *)featuresWithResource:(NSString *)resource {
  NSString *path = [[NSBundle bundleForClass:[self class]] pathForResource:resource
                                                                    ofType:@"geojson"];
  NSString *file = [[NSString alloc] initWithContentsOfFile:path
                                                   encoding:NSUTF8StringEncoding
                                                      error:nil];
  NSData *data = [file dataUsingEncoding:NSUTF8StringEncoding];
  GMUGeoJSONParser *parser = [[GMUGeoJSONParser alloc] initWithData:data];
  [parser parse];
  return parser.features;
}

- (void)testInitWithURL {
  NSString *path = [[NSBundle bundleForClass:[self class]] pathForResource:@"GeoJSON_Point_Test"
                                                                    ofType:@"geojson"];
  NSURL *url = [NSURL fileURLWithPath:path];
  GMUGeoJSONParser *parser = [[GMUGeoJSONParser alloc] initWithURL:url];
  [parser parse];
  XCTAssertEqual(parser.features.count, 1);
}

- (void)testInitWithStream {
  NSString *path = [[NSBundle bundleForClass:[self class]] pathForResource:@"GeoJSON_Point_Test"
                                                                    ofType:@"geojson"];
  NSString *file = [[NSString alloc] initWithContentsOfFile:path
                                                   encoding:NSUTF8StringEncoding
                                                      error:nil];
  NSData *data = [file dataUsingEncoding:NSUTF8StringEncoding];
  NSInputStream *stream = [[NSInputStream alloc] initWithData:data];
  GMUGeoJSONParser *parser = [[GMUGeoJSONParser alloc] initWithStream:stream];
  [parser parse];
  XCTAssertEqual(parser.features.count, 1);
}

- (void)testParsePoint {
  NSArray<GMUFeature *> *features = [self featuresWithResource:@"GeoJSON_Point_Test"];
  XCTAssertEqual(features.count, 1);
  GMUPoint *point = features.firstObject.geometry;
  XCTAssertEqual(point.coordinate.latitude, 0.5);
  XCTAssertEqual(point.coordinate.longitude, 102.0);
}

- (void)testParseLineString {
  NSArray<GMUFeature *> *features = [self featuresWithResource:@"GeoJSON_LineString_Test"];
  XCTAssertEqual(features.count, 1);
  GMSMutablePath *path = [[GMSMutablePath alloc] init];
  [path addLatitude:0.0 longitude:102.0];
  [path addLatitude:1.0 longitude:103.0];
  GMULineString *lineString = features.firstObject.geometry;
  XCTAssertEqualObjects(lineString.path.encodedPath, path.encodedPath);
}

- (void)testParsePolygon {
  NSArray<GMUFeature *> *features = [self featuresWithResource:@"GeoJSON_Polygon_Test"];
  XCTAssertEqual(features.count, 1);
  GMSMutablePath *outerPath = [[GMSMutablePath alloc] init];
  [outerPath addLatitude:10 longitude:10];
  [outerPath addLatitude:20 longitude:10];
  [outerPath addLatitude:20 longitude:20];
  [outerPath addLatitude:10 longitude:20];
  [outerPath addLatitude:10 longitude:10];
  GMSMutablePath *innerPath = [[GMSMutablePath alloc] init];
  [innerPath addLatitude:12.5 longitude:12.5];
  [innerPath addLatitude:17.5 longitude:12.5];
  [innerPath addLatitude:17.5 longitude:17.5];
  [innerPath addLatitude:12.5 longitude:17.5];
  [innerPath addLatitude:12.5 longitude:12.5];
  GMUPolygon *polygon = features.firstObject.geometry;
  XCTAssertEqualObjects(polygon.paths.firstObject.encodedPath, outerPath.encodedPath);
  XCTAssertEqualObjects(polygon.paths.lastObject.encodedPath, innerPath.encodedPath);
}

- (void)testParseMultiPoint {
  NSArray<GMUFeature *> *features = [self featuresWithResource:@"GeoJSON_MultiPoint_Test"];
  XCTAssertEqual(features.count, 1);
  GMUGeometryCollection *points = features.firstObject.geometry;
  GMUPoint *firstPoint = points.geometries.firstObject;
  GMUPoint *secondPoint = points.geometries.lastObject;
  XCTAssertEqual(firstPoint.coordinate.latitude, 0.0);
  XCTAssertEqual(firstPoint.coordinate.longitude, 100.0);
  XCTAssertEqual(secondPoint.coordinate.latitude, 1.0);
  XCTAssertEqual(secondPoint.coordinate.longitude, 101.0);
}

- (void)testParseMultiLineString {
  NSArray<GMUFeature *> *features = [self featuresWithResource:@"GeoJSON_MultiLineString_Test"];
  XCTAssertEqual(features.count, 1);
  GMUGeometryCollection *lineStrings = features.firstObject.geometry;
  GMULineString *firstLineString = lineStrings.geometries.firstObject;
  GMULineString *secondLineString = lineStrings.geometries.lastObject;
  GMSMutablePath *firstPath = [[GMSMutablePath alloc] init];
  [firstPath addLatitude:0.0 longitude:100.0];
  [firstPath addLatitude:1.0 longitude:101.0];
  GMSMutablePath *secondPath = [[GMSMutablePath alloc] init];
  [secondPath addLatitude:2.0 longitude:102.0];
  [secondPath addLatitude:3.0 longitude:103.0];
  XCTAssertEqualObjects(firstLineString.path.encodedPath, firstPath.encodedPath);
  XCTAssertEqualObjects(secondLineString.path.encodedPath, secondPath.encodedPath);
}

- (void)testParseMultiPolygon {
  NSArray<GMUFeature *> *features = [self featuresWithResource:@"GeoJSON_MultiPolygon_Test"];
  XCTAssertEqual(features.count, 1);
  GMUGeometryCollection *polygons = features.firstObject.geometry;
  GMUPolygon *firstPolygon = polygons.geometries.firstObject;
  GMUPolygon *secondPolygon = polygons.geometries.lastObject;
  GMSMutablePath *firstPath = [[GMSMutablePath alloc] init];
  [firstPath addLatitude:2.0 longitude:102.0];
  [firstPath addLatitude:2.0 longitude:103.0];
  [firstPath addLatitude:3.0 longitude:103.0];
  [firstPath addLatitude:3.0 longitude:102.0];
  [firstPath addLatitude:2.0 longitude:102.0];
  GMSMutablePath *secondPath = [[GMSMutablePath alloc] init];
  [secondPath addLatitude:0.0 longitude:100.0];
  [secondPath addLatitude:0.0 longitude:101.0];
  [secondPath addLatitude:1.0 longitude:101.0];
  [secondPath addLatitude:1.0 longitude:100.0];
  [secondPath addLatitude:0.0 longitude:100.0];
  XCTAssertEqualObjects(firstPolygon.paths.firstObject.encodedPath, firstPath.encodedPath);
  XCTAssertEqualObjects(secondPolygon.paths.firstObject.encodedPath, secondPath.encodedPath);
}

- (void)testParseGeometryCollection {
  NSArray<GMUFeature *> *features = [self featuresWithResource:@"GeoJSON_GeometryCollection_Test"];
  XCTAssertEqual(features.count, 1);
  GMUGeometryCollection *geometries = features.firstObject.geometry;
  GMUPoint *point = geometries.geometries.firstObject;
  GMULineString *lineString = geometries.geometries.lastObject;
  GMSMutablePath *path = [[GMSMutablePath alloc] init];
  [path addLatitude:0.0 longitude:101.0];
  [path addLatitude:1.0 longitude:102.0];
  XCTAssertEqual(point.coordinate.latitude, 0.0);
  XCTAssertEqual(point.coordinate.longitude, 100.0);
  XCTAssertEqualObjects(lineString.path.encodedPath, path.encodedPath);
}

- (void)testParseFeature {
  NSArray<GMUFeature *> *features = [self featuresWithResource:@"GeoJSON_Feature_Test"];
  XCTAssertEqual(features.count, 1);
  GMUFeature *feature = features.firstObject;
  GMUPoint *actualPoint = feature.geometry;
  CLLocationCoordinate2D northEast = CLLocationCoordinate2DMake(10, 10);
  CLLocationCoordinate2D southWest = CLLocationCoordinate2DMake(-10, -10);
  GMSCoordinateBounds *bounds = [[GMSCoordinateBounds alloc] initWithCoordinate:northEast
                                                                     coordinate:southWest];
  XCTAssertEqualObjects(feature.identifier, @"Test Feature");
  XCTAssertEqualObjects(feature.properties, @{ @"description" : @"A feature for unit testing" });
  XCTAssertEqualObjects(feature.boundingBox, bounds);
  XCTAssertEqual(actualPoint.coordinate.latitude, 0.5);
  XCTAssertEqual(actualPoint.coordinate.longitude, 102.0);
}

- (void)testParseFeatureCollection {
  NSArray<GMUFeature *> *features = [self featuresWithResource:@"GeoJSON_FeatureCollection_Test"];
  XCTAssertEqual(features.count, 2);
  GMUFeature *firstFeature = features.firstObject;
  GMUPoint *point = firstFeature.geometry;
  XCTAssertEqual(point.coordinate.latitude, 0.5);
  XCTAssertEqual(point.coordinate.longitude, 102.0);
  GMUFeature *secondFeature = features.lastObject;
  GMULineString *lineString = secondFeature.geometry;
  GMSMutablePath *path = [[GMSMutablePath alloc] init];
  [path addLatitude:0.0 longitude:102.0];
  [path addLatitude:1.0 longitude:103.0];
  XCTAssertEqualObjects(lineString.path.encodedPath, path.encodedPath);
}

@end

File: Tests/GoogleMapsUtilsSwiftTests/unit/Geometry/GMUGeoJSONParserTest.swift
// Copyright 2024 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import XCTest

@testable import GoogleMapsUtils

final class GMUGeoJSONParserTest: XCTestCase {

  // Helper function to load GeoJSON data
  private func features(withResource resource: String) -> [GMUFeature] {
    #if SWIFT_PACKAGE
    guard let path = Bundle.module.path(forResource: resource, ofType: "geojson"),
          let fileContents = try? String(contentsOfFile: path, encoding: .utf8),
          let data = fileContents.data(using: .utf8) else {
      XCTFail("GeoJSON resource not found or failed to load.")
      return []
    }
    #else
    let bundle = Bundle(for: Self.self)
    guard let path = bundle.path(forResource: resource, ofType: "geojson"),
          let fileContents = try? String(contentsOfFile: path, encoding: .utf8),
          let data = fileContents.data(using: .utf8) else {
      XCTFail("GeoJSON resource not found or failed to load.")
      return []
    }
    #endif

    let parser = GMUGeoJSONParser(data: data)
    parser.parse()
    return parser.features as! [GMUFeature]
  }

  func testInitWithURL() {
    #if SWIFT_PACKAGE
    guard let path = Bundle.module.path(forResource: "GeoJSON_Point_Test", ofType: "geojson") else {
      XCTFail("GeoJSON resource not found.")
      return
    }
    #else
    let bundle = Bundle(for: Self.self)
    guard let path = bundle.path(forResource: "GeoJSON_Point_Test", ofType: "geojson") else {
      XCTFail("GeoJSON resource not found.")
      return
    }
    #endif

    let url = URL(fileURLWithPath: path)
    let parser = GMUGeoJSONParser(url: url)
    parser.parse()
    XCTAssertEqual(parser.features.count, 1)
  }

  func testInitWithStream() {
    #if SWIFT_PACKAGE
    guard let path = Bundle.module.path(forResource: "GeoJSON_Point_Test", ofType: "geojson"),
          let fileContents = try? String(contentsOfFile: path, encoding: .utf8),
          let data = fileContents.data(using: .utf8) else {
      XCTFail("GeoJSON resource not found or failed to load.")
      return
    }
    #else
    let bundle = Bundle(for: Self.self)
    guard let path = bundle.path(forResource: "GeoJSON_Point_Test", ofType: "geojson"),
          let fileContents = try? String(contentsOfFile: path, encoding: .utf8),
          let data = fileContents.data(using: .utf8) else {
      XCTFail("GeoJSON resource not found or failed to load.")
      return
    }
    #endif

    let stream = InputStream(data: data)
    let parser = GMUGeoJSONParser(stream: stream)
    parser.parse()
    XCTAssertEqual(parser.features.count, 1)
  }

  func testParsePoint() {
    let features = features(withResource: "GeoJSON_Point_Test")
    XCTAssertEqual(features.count, 1)

    guard let point = features.first?.geometry as? GMUPoint else {
      XCTFail("Geometry is not a GMUPoint")
      return
    }

    XCTAssertEqual(point.coordinate.latitude, 0.5)
    XCTAssertEqual(point.coordinate.longitude, 102.0)
  }

  func testParseLineString() {
    let features = features(withResource: "GeoJSON_LineString_Test")
    XCTAssertEqual(features.count, 1)

    guard let lineString = features.first?.geometry as? GMULineString else {
      XCTFail("Geometry is not a GMULineString")
      return
    }

    let path = GMSMutablePath()
    path.addLatitude(0.0, longitude: 102.0)
    path.addLatitude(1.0, longitude: 103.0)

    XCTAssertEqual(lineString.path.encodedPath(), path.encodedPath())
  }

  func testParsePolygon() {
    let features = features(withResource: "GeoJSON_Polygon_Test")
    XCTAssertEqual(features.count, 1)

    guard let polygon = features.first?.geometry as? GMUPolygon else {
      XCTFail("Geometry is not a GMUPolygon")
      return
    }


    let outerPath = GMSMutablePath()
    outerPath.addLatitude(10, longitude: 10)
    outerPath.addLatitude(20, longitude: 10)
    outerPath.addLatitude(20, longitude: 20)
    outerPath.addLatitude(10, longitude: 20)
    outerPath.addLatitude(10, longitude: 10)

    let innerPath = GMSMutablePath()
    innerPath.addLatitude(12.5, longitude: 12.5)
    innerPath.addLatitude(17.5, longitude: 12.5)
    innerPath.addLatitude(17.5, longitude: 17.5)
    innerPath.addLatitude(12.5, longitude: 17.5)
    innerPath.addLatitude(12.5, longitude: 12.5)

    XCTAssertEqual(polygon.paths.first?.encodedPath(), outerPath.encodedPath())
    XCTAssertEqual(polygon.paths.last?.encodedPath(), innerPath.encodedPath())
  }

  func testParseMultiPoint() {
    let features = features(withResource: "GeoJSON_MultiPoint_Test")
    XCTAssertEqual(features.count, 1)

    guard let points = features.first?.geometry as? GMUGeometryCollection else {
      XCTFail("Geometry is not a GMUGeometryCollection")
      return
    }

    guard let firstPoint = points.geometries.first as? GMUPoint,
          let secondPoint = points.geometries.last as? GMUPoint else {
      XCTFail("Geometry is not a GMUPoint")
      return
    }

    XCTAssertEqual(firstPoint.coordinate.latitude, 0.0)
    XCTAssertEqual(firstPoint.coordinate.longitude, 100.0)
    XCTAssertEqual(secondPoint.coordinate.latitude, 1.0)
    XCTAssertEqual(secondPoint.coordinate.longitude, 101.0)
  }

  func testParseMultiLineString() {
    let features = features(withResource: "GeoJSON_MultiLineString_Test")
    XCTAssertEqual(features.count, 1)

    guard let lineStrings = features.first?.geometry as? GMUGeometryCollection else {
      XCTFail("Geometry is not a GMUGeometryCollection")
      return
    }

    let firstLineString = lineStrings.geometries.first as! GMULineString
    let secondLineString = lineStrings.geometries.last as! GMULineString

    let firstPath = GMSMutablePath()
    firstPath.addLatitude(0.0, longitude: 100.0)
    firstPath.addLatitude(1.0, longitude: 101.0)

    let secondPath = GMSMutablePath()
    secondPath.addLatitude(2.0, longitude: 102.0)
    secondPath.addLatitude(3.0, longitude: 103.0)

    XCTAssertEqual(firstLineString.path.encodedPath(), firstPath.encodedPath())
    XCTAssertEqual(secondLineString.path.encodedPath(), secondPath.encodedPath())
  }

  func testParseMultiPolygon() {
    let features = features(withResource: "GeoJSON_MultiPolygon_Test")
    XCTAssertEqual(features.count, 1)

    guard let polygons = features.first?.geometry as? GMUGeometryCollection else {
      XCTFail("Geometry is not a GMUPolygon")
      return
    }

    let firstPolygon = polygons.geometries.first as! GMUPolygon
    let secondPolygon = polygons.geometries.last as! GMUPolygon


    let firstPath = GMSMutablePath()
    firstPath.addLatitude(2.0, longitude: 102.0)
    firstPath.addLatitude(2.0, longitude: 103.0)
    firstPath.addLatitude(3.0, longitude: 103.0)
    firstPath.addLatitude(3.0, longitude: 102.0)
    firstPath.addLatitude(2.0, longitude: 102.0)

    let secondPath = GMSMutablePath()
    secondPath.addLatitude(0.0, longitude: 100.0)
    secondPath.addLatitude(0.0, longitude: 101.0)
    secondPath.addLatitude(1.0, longitude: 101.0)
    secondPath.addLatitude(1.0, longitude: 100.0)
    secondPath.addLatitude(0.0, longitude: 100.0)

    XCTAssertEqual(firstPolygon.paths.first?.encodedPath(), firstPath.encodedPath())
    XCTAssertEqual(secondPolygon.paths.last?.encodedPath(), secondPath.encodedPath())
  }

  func testParseGeomeryCollection() {
    let features = features(withResource: "GeoJSON_GeometryCollection_Test")
    XCTAssertEqual(features.count, 1)

    guard let collection = features.first?.geometry as? GMUGeometryCollection else {
      XCTFail("Geometry is not a GMUGeometryCollection")
      return
    }

    guard let point = collection.geometries.first as? GMUPoint else {
      XCTFail("Geometry is not a GMUPoint")
      return
    }

    guard let lineString = collection.geometries.last as? GMULineString else {
      XCTFail("Geometry is not a GMULineString")
      return
    }

    let path = GMSMutablePath()
    path.addLatitude(0.0, longitude: 101.0)
    path.addLatitude(1.0, longitude: 102.0)

    XCTAssertEqual(point.coordinate.latitude, 0.0)
    XCTAssertEqual(point.coordinate.longitude, 100.0)
    XCTAssertEqual(lineString.path.encodedPath(), path.encodedPath())
  }

  func testParseFeature() {
    let features = features(withResource: "GeoJSON_Feature_Test")
    XCTAssertEqual(features.count, 1)

    guard let feature = features.first else {
      XCTFail("No first feature")
      return
    }

    guard let actualPoint = feature.geometry as? GMUPoint else {
      XCTFail("Geometry is not a GMUPoint")
      return
    }

    let northEast: CLLocationCoordinate2D = CLLocationCoordinate2DMake(10, 10)
    let southWest: CLLocationCoordinate2D = CLLocationCoordinate2DMake(-10, -10)
    let bounds: GMSCoordinateBounds = GMSCoordinateBounds(coordinate: northEast, coordinate: southWest)
    let description: NSObject = "A feature for unit testing" as NSObject

    XCTAssertEqual(feature.identifier, "Test Feature")
    XCTAssertEqual(feature.properties, ["description": description])
    XCTAssertEqual(feature.boundingBox, bounds)
    XCTAssertEqual(actualPoint.coordinate.latitude, 0.5)
    XCTAssertEqual(actualPoint.coordinate.longitude, 102.0)
  }

  func testParseFeatureCollection() {
    let features = features(withResource: "GeoJSON_FeatureCollection_Test")
    XCTAssertEqual(features.count, 2)

    guard let firstFeature = features.first,
          let secondFeature = features.last else {
      XCTFail("Geometry is not a GMUFeature")
      return
    }

    let point = firstFeature.geometry as! GMUPoint
    let lineString = secondFeature.geometry as! GMULineString

    let path: GMSMutablePath = GMSMutablePath()
    path.addLatitude(0.0, longitude: 102.0)
    path.addLatitude(1.0, longitude: 103.0)

    XCTAssertEqual(point.coordinate.latitude, 0.5)
    XCTAssertEqual(point.coordinate.longitude, 102.0)
    XCTAssertEqual(lineString.path.encodedPath(), path.encodedPath())
  }

}
